### PR TITLE
Input nodes

### LINKED ISSUE #10696: Add inputs to grouped components
In the same way that we have an output component inside a grouped component we should also have inputs

image

image

These inputs cannot be deleted or edited.
They are autogenerated from the collapsed function's signature.

### PR DESCRIPTION
**Pull request by vitvakatu**
### Pull Request Description

Closes #10696 

https://github.com/user-attachments/assets/1163a288-7b91-462d-9774-beaf4fc34416

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- [x] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      [TypeScript](https://github.com/enso-org/enso/blob/develop/docs/style-guide/typescript.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- [x] Unit tests have been written where possible.

### CODE REVIEW

File: app/gui2/e2e/collapsingAndEntering.spec.ts [hunk 1]
```diff
@@ -124,6 +123,23 @@ test('Collapsing nodes', async ({ page }) => {
   await expect(locate.graphNodeByBinding(page, 'ten')).toExist()
 })
 
+test('Input node', async ({ page }) => {
+  await actions.goToGraph(page)
+  await enterToFunc2(page)
+
+  const inputNode = locate.inputNode(page)
+  await expect(inputNode).toHaveCount(1)
+  // Output node with identifier should have the icon and an identifier
```
Comment (farmaazon) [inline app/gui2/e2e/collapsingAndEntering.spec.ts]:
```suggestion
  // Input node with identifier should have the icon and an identifier
```

File: app/gui2/src/stores/graph/__tests__/graphDatabase.test.ts [hunk 2]
```diff
@@ -94,6 +95,7 @@ test('Reading graph from definition', () => {
 
   // Commented the connection from input node, as we don't support them yet.
   expect(Array.from(db.connections.allForward(), ([key]) => key)).toEqual([
+    id('parameter'),
     id('node1Id'),
     id('node2Id'),
   ])
```
Comment (farmaazon) [inline app/gui2/src/stores/graph/__tests__/graphDatabase.test.ts]: Please uncomment the comented expects below and remove the comment here.

File: app/gui2/src/stores/graph/graphDatabase.ts [hunk 3]
```diff
@@ -524,16 +559,18 @@ export class GraphDb {
   }
 }
 
-interface BlockLine {
+interface NodeInfo {
   outerAst: Ast.Ast
-  isLastInBlock: boolean
+  isOutput: boolean
+  isInput: boolean
+  argIndex: number | undefined
```
Comment (farmaazon) [inline app/gui2/src/stores/graph/graphDatabase.ts]: It's not clear what is the difference between NodeInfo and Node. Maybe this should be "NodeSource"? (A line or parameter).

File: app/gui2/src/stores/graph/graphDatabase.ts [hunk 4]
```diff
@@ -160,9 +160,6 @@ export class GraphDb {
 
   connections = new ReactiveIndex(this.bindings.bindings, (alias, info) => {
     const srcNode = this.getPatternExpressionNodeId(alias)
-    // Display connection starting from existing node.
-    //TODO[ao]: When implementing input nodes, they should be taken into account here.
-    if (srcNode == null) return []
```
Comment (farmaazon) [inline app/gui2/src/stores/graph/graphDatabase.ts]: I'm not sure about this change. What about nodes with lambdas? We should omit the "connection" from lambda parameter to its usage.

**Review by farmaazon — COMMENTED** [review]
Looking at it in action, now I think `undefined` would be clearer...

File: app/gui2/src/stores/graph/graphDatabase.ts [hunk 5]
```diff
@@ -336,23 +333,51 @@ export class GraphDb {
...
-        const data = shallowReactive({ isLastInBlock, outerAst })
+        const data = shallowReactive({ isOutput, outerAst: ast, isInput, argIndex })
         const stop = watchEffect(() =>
-          this.updateNodeStructure(nodeId, data.outerAst, data.isLastInBlock),
+          this.updateNodeStructure(
+            nodeId,
+            data.outerAst,
+            data.isOutput,
+            data.isInput,
+            data.argIndex,
```
Comment (kazcw) [inline app/gui2/src/stores/graph/graphDatabase.ts:352]: `updateNodeStructure` doesn't need to set `argIndex` if `updateNodes` does. The purpose of the separation is `updateNodeStructure` scans deep inside the node for properties derived from its internal AST, while `updateNodes` has to scan the whole function, but only shallowly, so it is only invalidated by high-level changes. In this case we have the `argIndex` data already without looking deeper into the AST.

File: app/gui2/src/stores/graph/graphDatabase.ts [hunk 6]
```diff
@@ -567,6 +604,8 @@ export interface NodeDataFromAst {
   conditionalPorts: Set<Ast.AstId>
   /** An AST node containing the node's documentation comment. */
   docs: Ast.Documented | undefined
+  /** The index of the argument in the function's argument list, if the node is an input node. */
+  argIndex: number | undefined
```
Comment (kazcw) [inline app/gui2/src/stores/graph/graphDatabase.ts:608]: This representation seems a bit "denormalized"--almost all the other fields here are not applicable to an input node

**Comment by kazcw** [comment]
I made a change to be able to handle all cases of complex argument definitions. The parser had the information, we were just lacking the `Ast` wrappers.

File: app/gui2/src/stores/graph/__tests__/graphDatabase.test.ts (hunk 2 above)
Comment (vitvakatu) [reply app/gui2/src/stores/graph/__tests__/graphDatabase.test.ts]: Done

File: app/gui2/src/stores/graph/graphDatabase.ts (hunk 3 above)
Comment (vitvakatu) [reply app/gui2/src/stores/graph/graphDatabase.ts]: True, thank you for suggestion, I was struggling with finding a better naming.

File: app/gui2/src/stores/graph/graphDatabase.ts (hunk 4 above)
Comment (vitvakatu) [reply app/gui2/src/stores/graph/graphDatabase.ts]: True, fixed it.

File: app/gui2/src/stores/graph/graphDatabase.ts [hunk 7]
```diff
@@ -559,10 +546,17 @@ export class GraphDb {
   }
 }
 
-interface NodeInfo {
+/** Source code data of the specific node. */
+interface NodeSource {
+  /** The outer AST of the node (see {@link NodeDataFromAst.outerExpr}). */
```
Comment (farmaazon) [inline app/gui2/src/stores/graph/graphDatabase.ts:551]: We could add information to `NodeDataFromAst.outerExpr` what it is for input node (it is only arg name? Or the name with default value and type?)

**Review by farmaazon — APPROVED** [review]

**Pull request merged**